File: pkgs/nixos-boot-cfg/nixos-boot-cfg.py
#!/usr/bin/env python3
"""NIXOS_BOOT_CFG - Mobile NixOS generation selection storage

Layout: 1MB misc partition divided into two 512KB blocks
  Block 0 (primary): offset 0, contains config at start
  Block 1 (backup): offset 524288, contains config at start

Format (75 bytes):
  0-13:   Magic "NIXOS_BOOT_CFG"
  14:     Null byte
  15:     Version (1)
  16:     Flags
  17-20:  Reserved
  21-74:  Path (null-terminated, max 53 chars)
"""

import sys
import os
import struct

MISC_PART = "/dev/disk/by-partlabel/misc"
BLOCK_SIZE = 524288  # 512KB
CFG_SIZE = 75


class BootConfig:
    def __init__(self, flags=0, path=""):
        self.version = 1
        self.flags = flags
        self.path = path[:53]  # Max 53 chars

    def to_bytes(self):
        """Serialize to 75 bytes"""
        data = b"NIXOS_BOOT_CFG\x00"  # 15 bytes
        data += struct.pack("B", self.version)  # 1 byte
        data += struct.pack("B", self.flags)  # 1 byte
        data += b"\x00\x00\x00\x00"  # 4 bytes reserved
        path_bytes = self.path.encode("utf-8")
        data += path_bytes  # path
        data += b"\x00" * (54 - len(path_bytes))  # padding
        return data

    @classmethod
    def from_bytes(cls, data):
        """Deserialize from bytes"""
        if len(data) < 22:
            return None
        if data[:15] != b"NIXOS_BOOT_CFG\x00":
            return None
        if data[15] != 1:
            return None

        config = cls()
        config.version = data[15]
        config.flags = data[16]
        # Extract path up to first null
        path_bytes = data[21:75]
        null_idx = path_bytes.find(b"\x00")
        if null_idx >= 0:
            path_bytes = path_bytes[:null_idx]
        config.path = path_bytes.decode("utf-8", errors="replace")
        return config


def read_boot_cfg(misc_path=MISC_PART):
    """Read config, trying primary then backup"""
    for block_num in [0, 1]:
        offset = block_num * BLOCK_SIZE
        try:
            with open(misc_path, "rb") as f:
                f.seek(offset)
                data = f.read(CFG_SIZE)
                if len(data) >= CFG_SIZE:
                    config = BootConfig.from_bytes(data)
                    if config:
                        return config, block_num
        except (IOError, OSError):
            pass
    return None, None


def write_boot_cfg(flags, path, misc_path=MISC_PART):
    """Write config to both blocks"""
    config = BootConfig(flags, path)
    data = config.to_bytes()

    try:
        # Write to backup first (block 1)
        with open(misc_path, "r+b") as f:
            f.seek(BLOCK_SIZE)
            f.write(data)

        # Then write to primary (block 0)
        with open(misc_path, "r+b") as f:
            f.seek(0)
            f.write(data)

        return True
    except (IOError, OSError) as e:
        print(f"Error writing: {e}", file=sys.stderr)
        return False


def clear_boot_cfg(misc_path=MISC_PART):
    """Clear both blocks"""
    try:
        zeros = b"\x00" * CFG_SIZE
        with open(misc_path, "r+b") as f:
            f.seek(0)
            f.write(zeros)
            f.seek(BLOCK_SIZE)
            f.write(zeros)
        return True
    except (IOError, OSError):
        return False


def main():
    if len(sys.argv) < 2:
        print("Usage: nixos-boot-cfg {read|write|clear} [args...]", file=sys.stderr)
        sys.exit(1)

    cmd = sys.argv[1]
    misc_path = MISC_PART if len(sys.argv) < 4 else sys.argv[-1]

    if cmd == "read":
        config, block = read_boot_cfg(misc_path)
        if config:
            print(f"{config.version} {config.flags} {config.path} {block}")
            sys.exit(0)
        else:
            sys.exit(1)

    elif cmd == "write":
        if len(sys.argv) < 4:
            print(
                "Usage: nixos-boot-cfg write <flags> <path> [misc_partition]",
                file=sys.stderr,
            )
            sys.exit(1)
        flags = int(sys.argv[2])
        path = sys.argv[3]
        if write_boot_cfg(flags, path, misc_path):
            sys.exit(0)
        else:
            sys.exit(1)

    elif cmd == "clear":
        if clear_boot_cfg(misc_path):
            sys.exit(0)
        else:
            sys.exit(1)

    else:
        print("Unknown command", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: pkgs/nixos-boot-cfg/nixos-boot-cfg.c
/* NIXOS_BOOT_CFG - Mobile NixOS generation selection storage
 * Layout: 1MB misc partition divided into two 512KB blocks
 *   Block 0 (primary): offset 0
 *   Block 1 (backup): offset 524288
 * Format (75 bytes):
 *   0-13:   Magic "NIXOS_BOOT_CFG"
 *   14:     Null byte
 *   15:     Version (1)
 *   16:     Flags
 *   17-20:  Reserved
 *   21-74:  Path (null-terminated, max 53 chars)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

#define MISC_PART_DEFAULT "/dev/disk/by-partlabel/misc"
#define BLOCK_SIZE 524288  /* 512KB */
#define CFG_SIZE 75
#define PATH_MAX 53

struct boot_config {
    uint8_t version;
    uint8_t flags;
    char path[PATH_MAX + 1];  /* +1 for null */
};

static const char MAGIC[] = "NIXOS_BOOT_CFG";

/* Read config from a specific block */
int read_block(int fd, int block_num, struct boot_config *cfg) {
    uint8_t buf[CFG_SIZE];
    off_t offset = (off_t)block_num * BLOCK_SIZE;
    
    if (lseek(fd, offset, SEEK_SET) != offset) {
        return -1;
    }
    
    ssize_t n = read(fd, buf, CFG_SIZE);
    if (n != CFG_SIZE) {
        return -1;
    }
    
    /* Check magic (bytes 0-14) */
    if (memcmp(buf, MAGIC, 14) != 0 || buf[14] != 0) {
        return -1;
    }
    
    /* Check version */
    if (buf[15] != 1) {
        return -1;
    }
    
    cfg->version = buf[15];
    cfg->flags = buf[16];
    
    /* Extract path (bytes 21-74) */
    memset(cfg->path, 0, sizeof(cfg->path));
    int path_len = 0;
    for (int i = 21; i < CFG_SIZE && path_len < PATH_MAX; i++) {
        if (buf[i] == 0) break;
        cfg->path[path_len++] = buf[i];
    }
    cfg->path[path_len] = '\0';
    
    return 0;
}

/* Read config, trying primary then backup */
int read_boot_cfg(const char *misc_path, struct boot_config *cfg, int *block_used) {
    int fd = open(misc_path, O_RDONLY);
    if (fd < 0) {
        return -1;
    }
    
    /* Try primary block (0) */
    if (read_block(fd, 0, cfg) == 0) {
        *block_used = 0;
        close(fd);
        return 0;
    }
    
    /* Try backup block (1) */
    if (read_block(fd, 1, cfg) == 0) {
        *block_used = 1;
        close(fd);
        return 0;
    }
    
    close(fd);
    return -1;
}

/* Write config to a block */
int write_block(int fd, int block_num, const struct boot_config *cfg) {
    uint8_t buf[CFG_SIZE];
    off_t offset = (off_t)block_num * BLOCK_SIZE;
    
    /* Clear buffer */
    memset(buf, 0, CFG_SIZE);
    
    /* Write magic */
    memcpy(buf, MAGIC, 14);
    buf[14] = 0;  /* null terminator */
    
    /* Write version and flags */
    buf[15] = cfg->version;
    buf[16] = cfg->flags;
    
    /* Reserved bytes 17-20 are already 0 */
    
    /* Write path (bytes 21-74) */
    int path_len = strlen(cfg->path);
    if (path_len > PATH_MAX) path_len = PATH_MAX;
    memcpy(buf + 21, cfg->path, path_len);
    
    if (lseek(fd, offset, SEEK_SET) != offset) {
        return -1;
    }
    
    ssize_t n = write(fd, buf, CFG_SIZE);
    if (n != CFG_SIZE) {
        return -1;
    }
    
    return 0;
}

/* Write config to both blocks */
int write_boot_cfg(const char *misc_path, const struct boot_config *cfg) {
    int fd = open(misc_path, O_RDWR);
    if (fd < 0) {
        fprintf(stderr, "Error opening %s: %s\n", misc_path, strerror(errno));
        return -1;
    }
    
    /* Write to backup first (block 1) */
    if (write_block(fd, 1, cfg) != 0) {
        close(fd);
        fprintf(stderr, "Error writing backup block\n");
        return -1;
    }
    
    /* Then write to primary (block 0) */
    if (write_block(fd, 0, cfg) != 0) {
        close(fd);
        fprintf(stderr, "Error writing primary block\n");
        return -1;
    }
    
    close(fd);
    return 0;
}

/* Clear both blocks */
int clear_boot_cfg(const char *misc_path) {
    int fd = open(misc_path, O_RDWR);
    if (fd < 0) {
        return -1;
    }
    
    uint8_t zeros[CFG_SIZE];
    memset(zeros, 0, CFG_SIZE);
    
    /* Clear primary */
    lseek(fd, 0, SEEK_SET);
    write(fd, zeros, CFG_SIZE);
    
    /* Clear backup */
    lseek(fd, BLOCK_SIZE, SEEK_SET);
    write(fd, zeros, CFG_SIZE);
    
    close(fd);
    return 0;
}

void usage(const char *prog) {
    fprintf(stderr, "Usage: %s {read|write|clear} [args...]\n", prog);
    fprintf(stderr, "  read [misc_partition]\n");
    fprintf(stderr, "  write <flags> <path> [misc_partition]\n");
    fprintf(stderr, "  clear [misc_partition]\n");
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        usage(argv[0]);
        return 1;
    }
    
    const char *cmd = argv[1];
    
    if (strcmp(cmd, "read") == 0) {
        const char *misc_path = (argc > 2) ? argv[2] : MISC_PART_DEFAULT;
        struct boot_config cfg;
        int block_used;
        
        if (read_boot_cfg(misc_path, &cfg, &block_used) != 0) {
            return 1;
        }
        
        printf("%d %d %s %d\n", cfg.version, cfg.flags, cfg.path, block_used);
        return 0;
    }
    else if (strcmp(cmd, "write") == 0) {
        if (argc < 4) {
            usage(argv[0]);
            return 1;
        }
        
        struct boot_config cfg;
        cfg.version = 1;
        cfg.flags = atoi(argv[2]);
        strncpy(cfg.path, argv[3], PATH_MAX);
        cfg.path[PATH_MAX] = '\0';
        
        const char *misc_path = (argc > 4) ? argv[4] : MISC_PART_DEFAULT;
        
        if (write_boot_cfg(misc_path, &cfg) != 0) {
            return 1;
        }
        return 0;
    }
    else if (strcmp(cmd, "clear") == 0) {
        const char *misc_path = (argc > 2) ? argv[2] : MISC_PART_DEFAULT;
        
        if (clear_boot_cfg(misc_path) != 0) {
            return 1;
        }
        return 0;
    }
    else {
        usage(argv[0]);
        return 1;
    }
}
